File: quickSort.cpp
#include <bits/stdc++.h>
using namespace std;

int partition(int arr[], int l, int h)
{
    int i = l - 1;
    int p = arr[h];
    for (int j = l; j < h; j++)
    {
        if (arr[j] < p)
        {
            i++;
            swap(arr[i], arr[j]);
        }
    }
    swap(arr[i + 1], arr[h]);
    return i + 1;
}

void quickSort(int arr[], int l, int h)
{
    if (l < h)
    {
        int p = partition(arr, l, h);
        quickSort(arr, l, p - 1);
        quickSort(arr, p + 1, h);
    }
}

int main()
{
    int a[] = {8, 6, 3, 2, 1};
    quickSort(a, 0, 4);
    for (int i = 0; i < 5; i++)
    {
        cout << a[i] << " ";
    }
    return 0;
}

File: quick_sort.py
def partition(arr, l, h):
    pivot = arr[h]
    i = l
    for j in range(l, h):
        if arr[j] <= pivot:
            arr[i], arr[j] = arr[j], arr[i]
            i += 1
    arr[i], arr[h] = arr[h], arr[i]
    return i


def quick_sort(arr, l, h):
    if l < h:
        pi = partition(arr, l, h)
        quick_sort(arr, l, pi-1)
        quick_sort(arr, pi+1, h)


data = [-2, 45, 0, 11, -9]
quick_sort(data, 0, len(data)-1)
print(data)


# Time Complexity: O(nlogn)
# Space Complexity: O(logn)
